gp-okta.py

```diff
@@ -385,67 +559,202 @@ def okta_redirect(conf, s, session_token, redirect_url):
+	with io.open(args.conf_file, 'rb') as fp:
+		config_contents = fp.read()
+
+	if args.conf_file.endswith('.gpg') and not args.gpg_decrypt:
+		err('conf file looks like gpg encrypted. Did you forget the --gpg-decrypt?')
+
+	if args.gpg_decrypt:
+		if not have_gnupg:
+			err('Need gnupg package for reading gnupg encrypted files. Consider doing `pip install python-gnupg` (or similar)')
+		gpg = gnupg.GPG(gnupghome=args.gpg_home)
+		decrypted_contents = gpg.decrypt(config_contents)
+
+		if not decrypted_contents.ok:
+			print('[ERROR] failed to decrypt config file:', file=sys.stderr)
+			print('[ERROR]     status: {}'.format(decrypted_contents.status), file=sys.stderr)
+			print('[ERROR]     error: {}'.format(decrypted_contents.stderr), file=sys.stderr)
+			sys.exit(1)
+
+		config_contents = decrypted_contents.data
+
+	conf = load_conf(config_contents)
+
 	s = requests.Session()
 	s.headers['User-Agent'] = 'PAN GlobalProtect'
-	saml_xml = paloalto_prelogin(conf, s)
+
+	if conf.get('client_cert'):
+		s.cert = conf.get('client_cert')
+
+	if args.show_list_of_gateways:
+		userauthcookie, gateways = paloalto_getconfig(conf, s)
+		print("Gateways:")
+		for gateway in sorted(gateways):
+			print(" ", gateway)
+		sys.exit(0)
+
+	another_dance = conf.get('another_dance', '').lower() in ['1', 'true']
+
+	if conf.get('gateway'):
+		gateway = conf.get('gateway').strip()
+	else:
+		gateway = None
+
+	if gateway and not another_dance:
+		vpn_url = 'https://{0}'.format(gateway)
+		if vpn_url != conf.get('vpn_url'):
+			log('Discarding \'vpn_url\', as concrete gateway is given and another_dance = 0')
+			conf['vpn_url'] = vpn_url
+
+	userauthcookie = None
+
+	if another_dance or not gateway:
+		saml_xml = paloalto_prelogin(conf, s)
+	else:
+		saml_xml = paloalto_prelogin(conf, s, gateway)
 	redirect_url = okta_saml(conf, s, saml_xml)
 	token = okta_auth(conf, s)
 	log('sessionToken: {0}'.format(token))
 	saml_username, prelogin_cookie = okta_redirect(conf, s, token, redirect_url)
+	if another_dance or not gateway:
+		userauthcookie, gateways = paloalto_getconfig(conf, s, saml_username, prelogin_cookie)
+		if not gateway and len(gateways):
+			gateway = gateways.pop() # this just grabs an arbitrary gateway
+
+	log('portal-userauthcookie: {0}'.format(userauthcookie))
+	log('gateway: {0}'.format(gateway))
 	log('saml-username: {0}'.format(saml_username))
 	log('prelogin-cookie: {0}'.format(prelogin_cookie))
-	userauthcookie = paloalto_getconfig(conf, s, saml_username, prelogin_cookie)
-	log('portal-userauthcookie: {0}'.format(userauthcookie))
-	
+
+	if another_dance:
+		# 1st step: dance with the portal, 2nd step: dance with the gateway
+		saml_xml = paloalto_prelogin(conf, s, gateway)
+		saml_username, prelogin_cookie = okta_saml_2(conf, s, gateway, saml_xml)
+		log('saml-username (2): {0}'.format(saml_username))
+		log('prelogin-cookie (2): {0}'.format(prelogin_cookie))
+
+	if (not userauthcookie or userauthcookie == 'empty') and prelogin_cookie != 'empty':
```

Comment:
Small request: for some reason my GP server works with the prelogin_cookie but not the userauthcookie. Can the token to use be an option that is set in the config?

Reply:
Or some preference setting? Could you please send a patch that is working on your side, then I can test here as well.

Reply:
sure I will polish it up a bit tomorrow and send it. sorry to hijack your PR but this is the smoothest experience for using GP out of the few options there are :)

Reply:
I was wrong, for some reason it wasn't working yesterday but it is working fine now.

The next thing is that my server requires a HIP report, so I need to be able to pass --csd-wrapper=hipreport.sh to openconnect at the end.

Reply:
(FYI I woke up and realized I can add that to the openconnect command in the config file... maybe something to consider adding to the README)